default/python/AI/pathfinding.py

```diff
@@ -0,0 +1,139 @@
+
+    # We have 1 free jump from supplied system into unsupplied systems.
+    # Thus, the target system must be at most maxFuel + 1 jumps away
+    # in order to reach the system under standard conditions.
+    # In some edge cases, we may have more supply here than what the
+    # supply graph suggests. For example, we could have recently refueled
+    # in a system that is now blockaded by an enemy or we have found
+    # the refuel special.
+    target_distance_from_supply = -min(state.get_system_supply(target), 0)
+    if (fleet.maxFuel + 1 < (target_distance_from_supply + minimum_fuel_at_target) and
+            universe.jumpDistance(start, target) > (start_fuel - minimum_fuel_at_target)):
+        # can't possibly reach this system with the required fuel
+        return None
+
+    # initialize data structures
+    path_cache = {}
+    queue = []
+
+    # add starting system to queue
+    heappush(queue, (shortest_possible_path_distance, path_information(distance=0, fuel=start_fuel, path=(start,))))
+
+    while queue:
+        # get next system u with path information
+        (_, path_info) = heappop(queue)
+        current = path_info.path[-1]
+
+        # did we reach the target?
+        if current == target:
+            # do we satisfy fuel constraints?
+            if path_info.fuel < minimum_fuel_at_target:
+                continue
+
+            return path_info
+
+        # add information about how we reached here to the cache
+        path_cache.setdefault(current, []).append(path_info)
+
+        # check if we have enough fuel to travel to neighbors
+        if path_info.fuel < 1:
+            continue
+
+        # add neighboring systems to the queue if the resulting path
+        # is either shorter or offers more fuel than the other paths
+        # which we already found to those systems
+        for neighbor in universe.getImmediateNeighbors(current, empire_id):
+            new_dist = path_info.distance + universe.linearDistance(current, neighbor)
+            new_fuel = (fleet.maxFuel if (neighbor in supplied_systems or current in supplied_systems) else
+                        path_info.fuel - 1)
+
+            # check if the node is already closed, i.e. a path was already found which both is shorter and offers
+            # more fuel. Priority queueing should ensure that all previously found paths here are shorter but check
+            # it anyway to be sure...
+            if any((dist <= new_dist and fuel >= new_fuel) for dist, fuel, _ in path_cache.get(neighbor, [])):
+                continue
+
+            # calculate the new distance prediction, i.e. the A* heuristic.
+            predicted_distance = new_dist + universe.shortestPathDistance(neighbor, target)
+
+            # Ignore paths that are much longer than the shortest possible path
+            if predicted_distance > max(2*shortest_possible_path_distance, shortest_possible_path_distance+2000):
```

Comment:
2000 is magic constant, pleas give it meaningful name, and may be some comments.